File: ios/flutter_meta_wearables_dat/Sources/flutter_meta_wearables_dat/StreamStateStreamHandler.swift
import Flutter
import MWDATCamera
import MWDATCore

/// Stream handler for stream state updates.
///
/// Set the `session` property when a `Stream` is created.
/// Clear it when the stream is torn down.
class StreamStateStreamHandler: NSObject, FlutterStreamHandler {
  /// The active stream to observe. Setting this property
  /// re-subscribes to the stream's state publisher.
  var session: MWDATCamera.Stream? {
    didSet { resubscribe() }
  }

  private var eventSink: FlutterEventSink?
  private var listenerToken: AnyListenerToken?

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    resubscribe()
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    cancelListener()
    return nil
  }

  /// Detaches from the current stream and pushes a terminal `stopped` to Dart.
  ///
  /// Plain `session = nil` detaches *silently*: `resubscribe()` cancels the
  /// listener and returns early on a nil session, so nothing is emitted. That is
  /// wrong for every plugin-initiated teardown, because those run on paths the
  /// SDK never reports on — the device-availability watchdog and a DeviceSession
  /// that stopped underneath us. Dart was left holding a live texture id and a
  /// `Texture` frozen on its last frame, with no state change and no error to
  /// act on. Emitting the terminal state here is what makes those teardowns
  /// observable.
  ///
  /// Safe to double-emit: if the SDK also reported `.stopped` before we
  /// detached, Dart sees the same terminal state twice.
  ///
  /// The emit is **synchronous** on purpose. The sole caller,
  /// `performTeardownStreamOnly()`, is `@MainActor` and calls this before its
  /// first suspension, so we are already on the platform thread and may touch
  /// the sink directly. Deferring it onto a `Task` instead would hand this
  /// terminal `stopped` an unordered slot: a restart that re-seeds the channel
  /// from `session`'s `didSet` could publish `starting`/`streaming` first, and
  /// the late `stopped` would then tell the app to drop a texture id that
  /// belongs to the *new* stream.
  @MainActor
  func detachEmittingStopped() {
    session = nil
    guard let events = eventSink else { return }
    events(Self.stoppedValue)
  }

  private func resubscribe() {
    cancelListener()
    guard let session = session, let events = eventSink else { return }

    // Send current state immediately
    events(Self.stateToInt(session.state))

    // Listen for future changes
    listenerToken = session.statePublisher.listen { state in
      Task { @MainActor in
        events(Self.stateToInt(state))
      }
    }
  }

  private func cancelListener() {
    if let token = listenerToken {
      Task { await token.cancel() }
      listenerToken = nil
    }
  }

  /// The Dart-facing value for `stopped`. Kept next to `stateToInt` so the two
  /// can't drift.
  private static let stoppedValue = 1

  /// Maps StreamState to the int values expected by Dart.
  private static func stateToInt(_ state: StreamState) -> Int {
    switch state {
    case .stopping:         return 0
    case .stopped:          return 1
    case .waitingForDevice: return 2
    case .starting:         return 3
    case .streaming:        return 4
    case .paused:           return 5
    @unknown default:       return 1 // default to stopped
    }
  }
}

File: android/src/main/kotlin/io/rodcone/flutter_meta_wearables_dat/StreamStateStreamHandler.kt
package io.rodcone.flutter_meta_wearables_dat

import com.meta.wearable.dat.camera.Stream
import com.meta.wearable.dat.camera.types.StreamState
import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Stream handler for stream session state updates from the DAT SDK.
 * Emits integer values matching the Dart `StreamState` enum.
 *
 * Set the [stream] property when a stream is added to a session.
 * Clear it when the stream is torn down.
 */
internal class StreamStateStreamHandler : EventChannel.StreamHandler {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var job: Job? = null
    private var eventSink: EventChannel.EventSink? = null

    /** The active stream to observe. Setting this resubscribes to the state flow. */
    var stream: Stream? = null
        set(value) {
            field = value
            resubscribe()
        }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        if (events == null) return
        eventSink = events
        resubscribe()
    }

    override fun onCancel(arguments: Any?) {
        job?.cancel()
        job = null
        eventSink = null
    }

    private fun resubscribe() {
        job?.cancel()
        job = null
        val sink = eventSink ?: return
        val currentStream = stream ?: return

        job =
                scope.launch {
                    currentStream.state.collect { state -> sink.success(mapState(state)) }
                }
    }

    fun dispose() {
        job?.cancel()
        job = null
        eventSink = null
        scope.cancel()
    }

    companion object {
        /**
         * Maps Android SDK StreamState to int values expected by Dart.
         *
         * Android enum: STARTING, STARTED, STREAMING, STOPPING, STOPPED, CLOSED
         * Dart enum:    stopping(0), stopped(1), waitingForDevice(2), starting(3), streaming(4), paused(5)
         *
         * Mapping notes:
         * - STARTED has no direct Dart equivalent; mapped to starting(3) since it precedes streaming.
         * - CLOSED has no direct Dart equivalent; mapped to stopped(1).
         * - Android has no WAITING_FOR_DEVICE or PAUSED states.
         */
        fun mapState(state: StreamState): Int {
            return when (state) {
                StreamState.STOPPING -> 0
                StreamState.STOPPED -> 1
                StreamState.STARTING -> 3
                StreamState.STARTED -> 3
                StreamState.STREAMING -> 4
                StreamState.CLOSED -> 1
            }
        }
    }
}
